### ISSUE TITLE
flutter_riverpod v2.1.3 broke my code, throws type 'AsyncData<void>' is not a subtype of type 'AsyncValue<int>'  exception

### ISSUE THREAD

**Issue by iLoveDocs** [rrousselGit/riverpod#2120]
**Describe the bug**
I can't assign `AsyncValue<T>` to `AsyncValue<void>`. 

**To Reproduce**

Generate the code for this provider and call `foo()` function. An exception will be thrown: 

> E/flutter (11354): [ERROR:flutter/runtime/dart_vm_initializer.cc(41)] Unhandled Exception: type 'AsyncData<void>' is not a subtype of type 'AsyncValue<int>' of 'previous'

```dart
@riverpod
class Foo extends AutoDisposeAsyncNotifier<void> {
  @override
  FutureOr<void> build() => null;

  Future<void> foo() async {
    final newState = await AsyncValue.guard(_baz);

    // Throws an exception but if you directly assign above result to `state` then there's no exception.
    state = newState;
  }

  Future<int> _baz() async => 0;
}
```

**Expected behavior**
There should be no such exception. This code was working fine with version `2.1.1`.

**Comment by rrousselGit** [comment]
There's nothing I can do here. It's more of a Dart limitation
There's no real bug in Riverpod here.


The problem can be boiled down to the following:

```dart
List<Object> list = <int>[];
list.add('string');
```

This code runs but throws.


The way this scenario happens in Riverpod is through the `copyWithPrevious`:


```dart
AsyncValue<T> newState;
_state = newState.copyWithPrevious(state);
```

When considering generics, there's nothing wrong. But if you unwrap the generics, what happens is the following:


```dart
AsyncValue<int> newState;
AsyncValue<void> state;

newState.copyWithPrevious(state);
```


This code doesn't compile as `copyWithPrevious` expects `AsyncValue<T>`. But due to generics, this problem cannot be highlighted at compilation and instead happens at runtime (like with the list sample above).


It's a known issue with Dart generics.

**Comment by iLoveDocs** [comment]
Thanks, I understood what the problem is. But is it possible to change Riverpod implementation to the one which was in `v2.1.1 ` because that seems to be working. 

If not, can you suggest me what should I rather do? I'm thinking of doing something like (too much boilerplate code)

```dart
Future<void> foo() async {
  final newState = await AsyncValue.guard(_baz);
  if (newState.hasError) {
    state = AsyncError(newState.error!, newState.stackTrace!);
  } else if (newState.hasValue) {
    state = AsyncData(null);
  } else {
    // If this is the case, there's something wrong with how we are doing things.
    throw Exception('State is loading.');
  }
}
```

**Comment by rrousselGit** [comment]
The behavior change is due to a bug that was fixed.
Reverting this would re-add the bug

It's not something which can be truly fixed.
To resolve this on your side, you shouldn't downcast the asyncValue

So do:

```dart
AsyncValue.guard<void>(...)
```

to avoid having an `AsyncValue<int>` being assigned to `AsyncValue<void>`

**Comment by iLoveDocs** [comment]
Well, you're the author, so I don't know if there's another way to solve that bug which introduced this problem. But from pure Dart point's of view, these two code should work the same way: 

```dart
final newState = await AsyncValue.guard(...);
state = newState;
```

and 

```dart
state = await AsyncValue.guard(...);
```

Seeing above two code doing different things just lowers my confidence level using Riverpod. Even if I use `AsyncValue.guard<void>`, I can still assign it to `AsyncValue<int>` and compiler doesn't complain, but again this is a runtime exception. 

```dart
final newState = await AsyncValue.guard<int>(...);
final AsyncValue<void> state = newState; // No compile time error. 
```

**Comment by rrousselGit** [comment]
These two codes don't work the same way. They have a different type inference.

Doing:

```dart
final newState = await AsyncValue.guard(...);
```

have the AsyncValue type infer to whatever the callback of `guard` returns.

But doing:
```dart
state = await AsyncValue.guard(...);
```

have the type inference use the type of `state`

**Comment by iLoveDocs** [comment]
Sorry, what I meant was, if Dart allows 

```
void a = 42;
```

do does the Riverpod should allow.

**Comment by rrousselGit** [comment]
I agree.
It's not a huge deal for me right now though. If you fancy it, you can try fixing it.

It'd likely involve replacing internal `copyWithPrevious` usages with the reserse mechanism. So instead of:

```dart
newState.copyWithPrevious(previousState);
```

we'd do:

```dart
previousState.copyWtihNext(nextState);
```

with the return value not changing.

**Comment by iLoveDocs** [comment]
Yes, sure. It's not that of a big deal for us. I'll stick to `2.1.1` for now. But I'll love to see it getting fixed in the coming updates. Thank you for your time 👍

**Comment by iLoveDocs** [comment]
@rrousselGit  Hi, could you please tell me if you have plans to fix this issue in the near future, I am unable to upgrade to the latest version as that will break my existing code. 

Thank you.

**Comment by rrousselGit** [comment]
it's not critical to me. You seem to be the only one with this problem

If you really need it, feel free to make a PR

**Comment by iLoveDocs** [comment]
Yes, I think so I'm the only one who is preventing code duplication like this:

```dart
Future<T?> _perform<T>(Future<T?> Function() operation) async {
  state = const AsyncLoading();
  final newState = await AsyncValue.guard(operation);
  if (mounted) state = newState;
  if (newState.hasValue) return newState.value;
  return null;
}
```

I never created any pull request and I also don't know where exactly I need to make changes in the source code which will fix this issue. So, I'll probably do the code duplication in my code.
